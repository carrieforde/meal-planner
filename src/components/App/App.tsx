import React from "react";
import { Route, Routes } from "react-router";
import { Catalog, MealPlans, Recipes, ShoppingList } from "views";

export const App: React.FC = () => (
  <Routes>
    <Route path="/" element={<ShoppingList />} />
    <Route path="shopping-list" element={<ShoppingList />} />
    <Route path="recipes" element={<Recipes />} />
    <Route path="meal-plans" element={<MealPlans />} />
    <Route path="catalog" element={<Catalog />} />
  </Routes>
);
