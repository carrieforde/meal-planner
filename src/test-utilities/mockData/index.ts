export * from "./mockedShoppingList";
export * from "./mockedCatalog";
