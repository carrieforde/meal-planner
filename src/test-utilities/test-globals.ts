const oldWindowLocation: string | undefined = undefined;

export function setNewLocation(url?: URL) {
  Reflect.deleteProperty(window, "location");
  window.location = url as unknown as Location;
  return window.location ?? oldWindowLocation;
}

export function mockGlobalVariable(key: string, value: unknown): void {
  const originalDescriptor = Object.getOwnPropertyDescriptor(window, key);

  Reflect.deleteProperty(window, key);

  beforeAll(() => {
    Object.defineProperty(window, key, {
      configurable: true,
      writable: true,
      value,
    });
  });

  afterAll(() => {
    Object.defineProperty(
      window,
      key,
      originalDescriptor as PropertyDescriptor
    );
  });
}
