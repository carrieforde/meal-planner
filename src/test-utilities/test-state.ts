import {
  initialDialogState,
  initialSnackbarState,
  useDialogState,
  useSnackbarState,
} from "store";

jest.mock("store", () => ({
  ...jest.requireActual("store"),
  useSnackbarState: jest.fn(() => ({
    isSnackbarOpen: false,
    message: null,
    severity: null,
  })),
  useDialogState: jest.fn(() => ({
    isDialogOpen: false,
    type: null,
  })),
}));

const mockedUseSnackbarState = jest.mocked(useSnackbarState);

export function useSnackbarStateMock(
  args?: Partial<ReturnType<typeof useSnackbarState>>
) {
  return mockedUseSnackbarState.mockReturnValue({
    ...initialSnackbarState,
    ...args,
  });
}

const mockedUseDialogState = jest.mocked(useDialogState);

export function useDialogStateMock(
  args?: Partial<ReturnType<typeof useDialogState>>
) {
  return mockedUseDialogState.mockReturnValue({
    ...initialDialogState,
    ...args,
  });
}
