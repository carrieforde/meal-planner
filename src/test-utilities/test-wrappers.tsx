import { FetchResult } from "@apollo/client";
import {
  MockedProvider,
  MockedResponse,
  ResultFunction,
} from "@apollo/client/testing";
import { DocumentNode, GraphQLError } from "graphql";
import { ReactNode } from "react";

export type TestApolloErrorType = "graphql" | "network";

export const testErrorMessage = "An error occurred!";

interface MockedWrapperArgs {
  result: FetchResult | ResultFunction<FetchResult>;
  query: DocumentNode;
  variables?: Record<string, any>;
}

export function createWrapper(
  args: MockedWrapperArgs[],
  errorType?: TestApolloErrorType
) {
  return function Wrapper({ children }: { children: ReactNode }) {
    const error =
      errorType === "network" ? new Error(testErrorMessage) : undefined;
    const errors =
      errorType === "graphql"
        ? [new GraphQLError(testErrorMessage)]
        : undefined;

    const mappedMocks: MockedResponse[] = args.map(
      ({ result, query, variables }) => {
        return {
          request: {
            query,
            variables,
          },
          result: {
            ...result,
            errors,
          },
          error,
        };
      }
    );

    return <MockedProvider mocks={mappedMocks}>{children}</MockedProvider>;
  };
}
