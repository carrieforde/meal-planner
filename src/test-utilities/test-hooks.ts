import { shoppingCategoriesMap, UnitMap } from "@constants";
import { useAddCatalogItem, useCatalogItems, useShoppingList } from "hooks";
import { mockGetShoppingList } from "./mockData";

jest.mock("hooks", () => ({
  useCatalogItems: jest.fn(() => ({
    data: undefined,
    error: undefined,
    loading: false,
    catalogRowData: [],
  })),
  useShoppingList: jest.fn(() => ({
    data: undefined,
    error: undefined,
    loading: false,
    listCategories: [],
    addItemToCart: jest.fn(),
    addItemToCartLoading: false,
  })),
  useAddCatalogItem: jest.fn(() => ({
    addCatalogItem: jest.fn(),
    loading: false,
  })),
}));

export const mockCatalogResponse = {
  catalog: [
    {
      id: "rSlQIDB2CZzvz4Klamug",
      name: "milk",
      category: "DAIRY",
      defaultUnit: "GALLON",
    },
    {
      id: "cVACp992BGJL3Pu5hNK7",
      name: "avocado",
      category: "PRODUCE",
      defaultUnit: null,
    },
    {
      id: "f0QbQE2UFnutNHJ29rbF",
      name: "bacon",
      category: "MEAT",
      defaultUnit: null,
    },
    {
      id: "7ron8dYBgH6zjBuYOpeT",
      name: "baking powder",
      category: "BAKING_SUPPLIES",
      defaultUnit: null,
    },
    {
      id: "1cBER9Mj8fBKW8BGtews",
      name: "baking soda",
      category: "BAKING_SUPPLIES",
      defaultUnit: null,
    },
  ],
};

export const mockMappedCatalogResponse = [
  {
    id: "rSlQIDB2CZzvz4Klamug",
    name: "milk",
    category: shoppingCategoriesMap.DAIRY,
    defaultUnit: UnitMap.GALLON,
  },
  {
    id: "cVACp992BGJL3Pu5hNK7",
    name: "avocado",
    category: shoppingCategoriesMap.PRODUCE,
    defaultUnit: null,
  },
  {
    id: "f0QbQE2UFnutNHJ29rbF",
    name: "bacon",
    category: shoppingCategoriesMap.MEAT,
    defaultUnit: null,
  },
  {
    id: "7ron8dYBgH6zjBuYOpeT",
    name: "baking powder",
    category: shoppingCategoriesMap.BAKING_SUPPLIES,
    defaultUnit: null,
  },
  {
    id: "1cBER9Mj8fBKW8BGtews",
    name: "baking soda",
    category: shoppingCategoriesMap.BAKING_SUPPLIES,
    defaultUnit: null,
  },
];

const mockedUseCatalogItems = jest.mocked(useCatalogItems);

const defaultUseCatalogItemsReturn: ReturnType<typeof useCatalogItems> = {
  data: mockCatalogResponse,
  error: undefined,
  loading: false,
  catalogRowData: mockMappedCatalogResponse,
};

export function useCatalogItemsMock(
  args?: Partial<ReturnType<typeof useCatalogItems>>
) {
  return mockedUseCatalogItems.mockReturnValue({
    ...defaultUseCatalogItemsReturn,
    ...args,
  });
}

const mockedUseShoppingList = jest.mocked(useShoppingList);

export const defaultUseShoppingListItemsReturn: ReturnType<
  typeof useShoppingList
> = {
  data: mockGetShoppingList.data,
  error: undefined,
  loading: false,
  listCategories: mockGetShoppingList.data.list.items
    .reduce(
      (acc: string[], item) =>
        acc.includes(item.category) ? acc : [...acc, item.category],
      []
    )
    .sort((a, b) => a.localeCompare(b)),
  addItemToCart: jest.fn(),
  addItemToCartLoading: false,
};

export function useShoppingListMock(
  args?: Partial<ReturnType<typeof useShoppingList>>
) {
  return mockedUseShoppingList.mockReturnValue({
    ...defaultUseShoppingListItemsReturn,
    ...args,
  });
}

const mockedUseAddCatalogItem = jest.mocked(useAddCatalogItem);

export const defaultUseAddCatalogItemReturn: ReturnType<
  typeof useAddCatalogItem
> = {
  addCatalogItem: jest.fn(),
  loading: false,
};

export function useAddCatalogItemMock(
  args?: Partial<ReturnType<typeof useAddCatalogItem>>
) {
  return mockedUseAddCatalogItem.mockReturnValue({
    ...defaultUseAddCatalogItemReturn,
    ...args,
  });
}
