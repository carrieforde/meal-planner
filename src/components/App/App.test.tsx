import { MockedProvider } from "@apollo/client/testing";
import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { useSnackbarStateMock } from "test-utilities/test-state";
import { App } from "./App";

const wrapper: React.FC = ({ children }) => (
  <MockedProvider>
    <BrowserRouter>{children}</BrowserRouter>
  </MockedProvider>
);

describe("App", () => {
  beforeEach(() => {
    useSnackbarStateMock();
  });

  it("should render the App component consistently", () => {
    const { container } = render(<App />, { wrapper });
    expect(container).toMatchSnapshot();
  });
});
