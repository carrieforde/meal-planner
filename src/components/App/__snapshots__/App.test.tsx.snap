// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render the App component consistently 1`] = `
<div>
  <button
    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
    tabindex="0"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="MenuIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
      />
    </svg>
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
  <main
    class="MuiBox-root css-1ourwbl"
  >
    <div
      class="MuiBox-root css-11od68z"
    >
      <span
        class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
        role="progressbar"
        style="width: 40px; height: 40px;"
      >
        <svg
          class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </span>
    </div>
  </main>
</div>
`;
